File: python/Luhn.py
import re
class Luhn:
    def __init__(self, card_num):
        self.card_num=re.sub(' ','',card_num)

    def valid(self):
        card_num=self.card_num
        if len(card_num)<=1 or not all(i.isnumeric() for i in card_num):
            return False 

        res=[int(i) for i in card_num]
        res[-2::-2] = [2*i - 9*(i>=5) for i in res[-2::-2]]

        return not (sum(res)%10)

File: c/Luhn.c
#include "luhn.h"
bool luhn(const char *num){
    if ((num[0] == '0') && !(num[1])){
        return false;
    }
    if (num[0] == ' ' && num[1] == '0')
        return false;
    
    int sum = 0;
    int flag = 0;
    int end = 0;
    while (num[end] != '\0')
        end++;
    
    for(int i = end-1; i >= 0 ; i--){
        if (num[i] != ' '){
            if (num[i] >= '0' && num[i] <= '9'){
                if(flag == 0){
                    sum += num[i] - '0';
                    flag = 1;
                } else {
                    int tmp = (num[i] - '0') * 2;
                    if (tmp > 9)
                        sum += (tmp - 9);
                    else
                        sum += tmp;
                    flag = 0;
                }
            }
            else
                return false;
        }
    }
    if(sum % 10 == 0)
        return true;
    return false;
}